src/utils/s3/s3_bulk_renaming.py

```diff
@@ -0,0 +1,130 @@
+import argparse
+import re
+import sys
+from typing import Optional
+
+import boto3  # type: ignore
+from botocore.exceptions import NoCredentialsError  # type: ignore
+
+
+def list_objects(bucket_name: str, prefix: Optional[str] = None) -> list[str]:
+    """
+    Lists all objects in the specified S3 bucket matching the given prefix.
+    Returns a list of object keys.
+    """
+    s3 = boto3.client("s3")
+    keys: list[str] = []
+
+    paginator = s3.get_paginator("list_objects_v2")
+    page_iterator = paginator.paginate(Bucket=bucket_name, Prefix=prefix)
+
+    try:
+        for page in page_iterator:
+            if "Contents" in page:
+                keys.extend(obj["Key"] for obj in page["Contents"])
+    except NoCredentialsError:
+        sys.exit("Error: Unable to locate credentials")
+    return keys
+
+
+def filter_objects(object_keys: list[str]) -> list[str]:
+    """
+    Filters the list of object keys based on the provided regex pattern.
+    Returns a list of filtered object keys.
+    """
+    return [key for key in object_keys if detect_timestamps_without_minutes(key)]
+
+
+def detect_timestamps_without_minutes(key: str) -> Optional[str]:
+    """
+    Returns the string if it contains a YYYY-MM-DDTHH timestamp.
+    Otherwise, return None
+    """
+    pattern_with_minutes = r"(\d{4}-\d{2}-\d{2}T\d{2}:\d{2})"  # "YYYY-MM-DDTHH:MM"
```

Comment:
Considering that this regex is used again in `update_timestamp`, I’d consider pulling it out and pre-compiling it at the top of the module into a global variable between your imports and the first function definition.

Reply:
Good point. Do you know if there are naming conventions for precompiled regexes?